### ISSUE TITLE
Endless file generation when dumping special tables like RSDP, DSDT, etc.

### ISSUE THREAD

**Issue by smwikipedia** [acpica/acpica#285]
[Symptom]:
I ported acpidump to UEFI Shell environment. When dumping special tables like RSDP, RSDT, XSDT, DSDT, FACS, acpidump.efi will generate too many files until it reaches AP_MAX_ACPI_FILES = 256.

This issue doesn't happen with normal tables like BERT, HEST. And doens't happen when dump table by index.

[Root cause]: 
These special tables are all singleton. But there's no necessary detection when dump them with instance number > 1.

**Comment by acpibob** [comment]
It does sound like a problem, but why are you using instance > 1?

**Comment by smwikipedia** [comment]
I am not using instance > 1. I just use the following command line:
acpidump.efi -b -n DSDT
And it will generate many dsdt.dat files. All identical.

The reason is, when use -n to dump a table by name, it will treat every talbe as potentially having multiple instances. It loops through 0 to AP_MAX_ACPI_FILES to call the AcpiOsGetTableByName(). For normal tables, there's a detection logic for excessive instnace number. For special tables like DSDT, there's not, so redundent files are generated.

**Comment by acpibob** [comment]
Here's what I see on win:

c:\temp>acpidump -b -n DSDT

 Directory of c:\temp

06/28/2017  08:44 AM    <DIR>          .
06/28/2017  08:44 AM    <DIR>          ..
06/28/2017  08:44 AM           126,458 dsdt.dat
06/28/2017  08:42 AM           734,476 dump.txt

I'll take a look at the code, but I'm not seeing any bad behavior. What version of acpidump are you using? here's the current version:

c:\temp>acpidump -v

Intel ACPI Component Architecture
ACPI Binary Table Dump Utility version 20170531
Copyright (c) 2000 - 2017 Intel Corporation

**Comment by smwikipedia** [comment]
Yes, on windows it's fine. The issue is with efi porting. The issue file is "\source\os_specific\efi\osefitbl.c". Function OslGetTable().

**Comment by acpibob** [comment]
Lv can have a look at this.

**Comment by acpibob** [comment]
Probably best if you open a bugzilla on this at

https://bugs.acpica.org

And point out that the problem is against the EFI version.
Thanks.

**Comment by smwikipedia** [comment]
I will contact him tomorrow.

**Comment by zetalog** [comment]
DSDT could have 2 instances, one for 32-bit OS, the other for 64-bit OS. That's why we allow to dump DSDT with instances.

**Comment by acpibob** [comment]
How many files are generated? 
It sounds like 256?

**Comment by zetalog** [comment]
Filed a bug and posted a fix here:

https://bugs.acpica.org/show_bug.cgi?id=1407

**Comment by zetalog** [comment]
> How many files are generated? It sounds like 256?
Yes, EFI support contains a very old bug.

Originally acpidump is developed to support 2 instances of DSDT/FACS but it's not tested and broken. On Linux we changed to support only 1 instance of DSDT/FACS when a bug was reported. But EFI support is not fixed.

Actually we should support 2 instances, as FACS/XFACS do contain different content. I tried to improve both Linux and EFI support.

**Comment by smwikipedia** [comment]
yes, from the logic, maximum 256 files can be generated.




发自网易邮箱大师
On 06/30/2017 08:58, zetalog wrote:

How many files are generated? It sounds like 256?
Yes, EFI support contains a very old bug.

Originally acpidump is developed to support 2 instances of DSDT/FACS but it's not tested and broken. On Linux we changed to support only 1 instance of DSDT/FACS when a bug was reported. But EFI support is not fixed.

Actually we should support 2 instances, as FACS/XFACS do contain different content. I tried to improve both Linux and EFI support.
